perception/detected_object_validation/src/object_lanelet_filter.cpp

```diff
@@ -0,0 +1,184 @@
+  const tf2_ros::Buffer & tf_buffer, const std::string & source_frame_id,
+  const std::string & target_frame_id, const rclcpp::Time & time)
+{
+  try {
+    geometry_msgs::msg::TransformStamped self_transform_stamped;
+    self_transform_stamped = tf_buffer.lookupTransform(
+      /*target*/ target_frame_id, /*src*/ source_frame_id, time,
+      rclcpp::Duration::from_seconds(0.5));
+    return self_transform_stamped.transform;
+  } catch (tf2::TransformException & ex) {
+    RCLCPP_WARN_STREAM(rclcpp::get_logger("object_lanelet_filter"), ex.what());
+    return boost::none;
+  }
+}
+namespace object_lanelet_filter
+{
+ObjectLaneletFilterNode::ObjectLaneletFilterNode(const rclcpp::NodeOptions & node_options)
+: Node("object_lanelet_filter_node", node_options),
+  tf_buffer_(this->get_clock()),
+  tf_listener_(tf_buffer_)
+{
+  using std::placeholders::_1;
+
+  // Set parameters
+  unknown_only_ = declare_parameter<bool>("unknown_only", false);
+
+  // Set publisher/subscriber
+  map_sub_ = this->create_subscription<autoware_auto_mapping_msgs::msg::HADMapBin>(
+    "input/vector_map", rclcpp::QoS{1}.transient_local(),
+    std::bind(&ObjectLaneletFilterNode::mapCallback, this, _1));
+  object_sub_ = this->create_subscription<autoware_auto_perception_msgs::msg::DetectedObjects>(
+    "input/object", rclcpp::QoS{1}, std::bind(&ObjectLaneletFilterNode::objectCallback, this, _1));
+  object_pub_ = this->create_publisher<autoware_auto_perception_msgs::msg::DetectedObjects>(
+    "output/object", rclcpp::QoS{1});
+}
+
+void ObjectLaneletFilterNode::mapCallback(
+  const autoware_auto_mapping_msgs::msg::HADMapBin::ConstSharedPtr map_msg)
+{
+  lanelet_map_ptr_ = std::make_shared<lanelet::LaneletMap>();
+  lanelet::utils::conversion::fromBinMsg(*map_msg, lanelet_map_ptr_);
+  const lanelet::ConstLanelets all_lanelets = lanelet::utils::query::laneletLayer(lanelet_map_ptr_);
+  road_lanelets_ = lanelet::utils::query::roadLanelets(all_lanelets);
+}
+
+void ObjectLaneletFilterNode::objectCallback(
+  const autoware_auto_perception_msgs::msg::DetectedObjects::ConstSharedPtr input_msg)
+{
+  using Label = autoware_auto_perception_msgs::msg::ObjectClassification;
+
+  // Guard
+  if (object_pub_->get_subscription_count() < 1) return;
+
+  autoware_auto_perception_msgs::msg::DetectedObjects output_object_msg;
+  output_object_msg.header = input_msg->header;
+
+  if (!lanelet_map_ptr_) {
+    RCLCPP_ERROR(get_logger(), "No vector map received.");
+    return;
+  }
+  autoware_auto_perception_msgs::msg::DetectedObjects transformed_objects;
+  if (!transformDetectedObjects(*input_msg, "map", tf_buffer_, transformed_objects)) {
+    RCLCPP_ERROR(get_logger(), "Failed transform to map.");
+    return;
+  }
+
+  // calculate convex hull
+  const auto convex_hull = getConvexHull(transformed_objects);
+  // get intersected lanelets
+  lanelet::ConstLanelets intersected_lanelets = getIntersectedLanelets(convex_hull, road_lanelets_);
```

Comment:
is it ok?
![image](https://user-images.githubusercontent.com/8327598/177443790-2e52b9af-1343-4aa7-8766-2949795b9bfa.png)

Reply:
Maybe considering every object's shape overlap  and convex hull of whole object is better?

Reply:
Pending

Reply:
I have tested software. This "getIntersectedLanelets" looks return not only lanelets with collisions but also inside of the convex hull.